### ISSUE TITLE
Cquery allpaths violating invariants?

### ISSUE THREAD

**Issue by cpsauer** [bazelbuild/bazel#18289]
### Description of the bug:

Hey wonderful Bazelers,

I was writing a feature for the (now reasonably popular) [compile commands extractor tool](https://github.com/hedronvision/bazel-compile-commands-extractor), and noticed some very strange behavior with allpaths/somepath in cquery and aquery, but not query, that I think prevents their use.

I'm seeing lots of instances where `cquery allpaths(A, C)` (reportedly) contains B, but `cquery somepath(B, C)` is empty. That seems like it should never happen!? If there were a path from A->C that contains B (A->B->C), then there ought to be a path from B->C, right?

---

More details:
- In all the examples I've seen so far: 
    There is indeed a path from A->B and no path from B->C. The problem is in `allpaths(A,C)`, which returns most but not all deps of A, including many extraneous targets.
   Query behaves properly; the problem is only in cquery/aquery.
- This is on the latest Bazel rolling (7.0.0-pre.20230420.2) and latest (6.1.2), latest macOS.

As always, thanks for all you do,
Chris
(ex-Googler)

**Comment by cpsauer** [comment]
okay, so here's a small, 100% repro, containing graphviz diagrams
[allpaths-test.zip](https://github.com/bazelbuild/bazel/files/11379289/allpaths-test.zip)

As above `bazel cquery 'allpaths(A, C)'` gives all kinds of erroneous stuff, including `B`
Declared graph structure is A->C->B.

Graphvis for that cquery confirms: The only in-edge to C is directly from A; there is no path from A -> C that goes through B. Despite that, many more nodes are returned. See allpaths-a-c-graph.png for visualization and allpaths-a-c-graph.in for graph. They were generated with `bazel cquery "allpaths(A, C)" --output graph > graph.in &&  dot -Tpng < graph.in > graph.png` per https://blog.bazel.build/2015/06/17/visualize-your-build.html
There are also graphs for `cquery deps(A)` for good measure.
The example does use ios_application (condensed from the original, sorry). But I was seeing the same with Android, too, if you want to try that instead.

**Comment by cpsauer** [comment]
For routing, I'm going to tag @gregestren, if that's ok, since that seems like a good guess. (Please coach me if not.)

**Comment by pcjanzen** [comment]
try `--noinclude_aspects`

**Comment by cpsauer** [comment]
(Also, heads that the order of allpaths/somepath seems to be reversed from what I'd have expected from the docs--a backtrace instead of a reversed post-order--not that I'm complaining for my use case.)

**Comment by cpsauer** [comment]
Wow, yeah, @pcjanzen, `--noinclude_aspects` disables whatever causes the issue. How'd you know? That's enough to unblock my use case, I think.

Still probably worth investigating and tracking down bc of the inconsistency?

We should probably also change "By default, cquery doesn't follow aspects" from https://bazel.build/query/cquery if the default has been thrown, as it sure seems to have.

**Comment by gregestren** [comment]
Thanks for the kind words, @cpsauer .

Without having dived into your example in detail (but thanks for outlining it), I suspect this is related to the [More accurate cquery](https://docs.google.com/document/d/1x-RqFFEcGIzVgtYHIq6s6Qt4vYtJWz3nuHIuqeI89dU/edit#heading=h.5mcn15i0e1ch) proposal I wrote up and would like to implement when I have the time. Precisely, as the title says, to make cquery more accurate, which is crucial to the point of the tool.

As far as I know, aspects and toolchain deps (which the doc covers) are the sole reason cquery could produce inaccurate outputs.

**Comment by gregestren** [comment]
Oh, about routing: if you give the issue a team label like https://github.com/bazelbuild/bazel/labels/team-Configurability, we do a [regular triage review](https://bazel.build/versions/6.0.0/contribute/maintainers-guide?authuser=0#label-own) to at least look at issues and move them from  https://github.com/bazelbuild/bazel/labels/untriaged to some priority label. It's not a consistent schedule but we aim to do this weekly.

Officially you don't even have to add a team label: a DevEx reviewer performs [initial triage](https://bazel.build/versions/6.0.0/contribute/maintainers-guide?authuser=0#lifecycle-issue) including assigning team labels. And they tend to be really on the ball schedule-wise. 

I'd still assign a team label if you know it because DevEx reviewers don't always know the right teams for the right issues.

**Comment by cpsauer** [comment]
Hey, Greg! Thanks for being great and for the cross-references.

You guys (obviously) know way more about the inner workings than I do, but it sure does look like an instance of the "results inaccurate ... allpaths" section to me.

Good for me to read about the new(ish) triage process (thanks!). I think normal users  can't assign labels without additional permissions, hence the tag. (Conversely, if that's too annoying, just lmk.)

Where shall we go from here? Sounds like you're already very well on it--but ok with leaving this open to track, if there isn't already a tracking issue? (There's also the docs-out-of-sync issue, above, and a bit more broadly for cquery/aquery.)

Cheers,
Chris

**Comment by gregestren** [comment]
Sure, happy to keep this open.

Can you elaborate more on the docs confusion? Which part of the docs is misleading?

**Comment by cpsauer** [comment]
I think it just needs a quick scan. I was seeing some markdown-type links that aren't getting converted up top, there's the aspects off-by-default text with aspects-on-by-default header, and it might be worth linking the design doc/this issue so people don't waste time? (There were also inconsistencies around ordering in the query docs, as mentioned above.)

**Comment by cpsauer** [comment]
^ Tossed up some quick PRs for the no brainers and broken links I'd seen. But I think for the others you've got the advantage of way more background knowledge.

**Comment by gregestren** [comment]
Duplicate of https://github.com/bazelbuild/bazel/issues/16310.
